lib/control/control-main.c

```diff
@@ -26,18 +26,61 @@
 #include "control-server.h"
 #include "control-commands.h"
 
-static ControlServer *control_server;
+#include <iv.h>
+#include <iv_event.h>
+
+typedef struct _ControlServerThread
+{
+  GThread *thread;
+  MainLoop *main_loop;
+  const gchar *control_name;
+  struct iv_event stop_requested;
+} ControlServerThread;
+
+static ControlServerThread control_server_thread;
+
+static void
+_thread_stop()
+{
+  iv_quit();
+}
+
+static void
+_register_stop_requested_event(ControlServerThread *self)
+{
+  IV_EVENT_INIT(&self->stop_requested);
+  self->stop_requested.handler = _thread_stop;
+  iv_event_register(&self->stop_requested);
+}
+
+static void
+_thread(gpointer user_data)
+{
+  ControlServerThread *cs_thread = (ControlServerThread *)user_data;
+  ControlServer *control_server = control_server_new(cs_thread->control_name,
+                                                     control_register_default_commands(cs_thread->main_loop));
+  iv_init();
+  {
+    _register_stop_requested_event(cs_thread);
+    control_server_start(control_server);
+    iv_main();
+    control_server_free(control_server);
+  }
+  iv_deinit();
+}
 
 void
-control_init(MainLoop *main_loop, const gchar *control_name)
+control_thread_start(MainLoop *main_loop, const gchar *control_name)
 {
-  control_server = control_server_new(control_name,
-                                      control_register_default_commands(main_loop));
-  control_server_start(control_server);
+  control_server_thread.main_loop = main_loop;
+  control_server_thread.control_name = control_name;
```

Comment:
Dont we need to ref the main loop instance?

Reply:
Only if the control server can run after mainloop is deinited.
Currently this is not possible (the thread is joinable and I call the join function in mainloop_deinit).
After I refactor the code(and move control server to main), it depends on when I call the mainloop_deinit.  I can extend Mainloop with a refctr logic.